assets/js/wpjm-stats.js

```diff
@@ -118,11 +108,134 @@ import { createHooks } from '@wordpress/hooks';
 		}, {} );
 
 		Object.keys( statsByTrigger ).forEach( function ( triggerName) {
-			hookStatsForTrigger( statsByTrigger, triggerName );
+			WPJMStats.hookStatsForTrigger( statsByTrigger, triggerName );
+		} );
+
+		WPJMStats.initCallbacks.forEach( function ( initCallback ) {
+			initCallback.call( null );
+		} );
+
+		WPJMStats.hooks.doAction( 'page-load' );
+	},
+
+	hookStatsForTrigger: function ( statsByTrigger, triggerName ) {
+		console.log( 'hookStatsForTrigger' );
+		const statsToRecord    = [];
+		const stats            = statsByTrigger[triggerName] || [];
+		const events           = {};
+		const statsByType      = {};
+
+		stats.forEach( function ( statToRecord ) {
+			if ( ! statsByType[statToRecord.type] ) {
+				statsByType[statToRecord.type] = [];
+			}
+
+			statsByType[statToRecord.type].push( statToRecord );
+			statsToRecord.push( statToRecord );
 		} );
 
-		// Kick things off.
-		console.log('kick thing off');
-		window.wpjmStatHooks.doAction( 'page-load' );
+		// Hook action to call logStats.
+		WPJMStats.hooks.addAction( triggerName, 'wpjm-stats', function () {
+			window.wpjmLogStats( statsToRecord );
+		}, 10 );
+
+		Object.keys( statsByType ).forEach( function ( type ) {
+			console.log( 'stats by type', type );
+			WPJMStats.types[type] && WPJMStats.types[type]( statsByType[type] );
+		} );
+	},
+
+	hooks: createHooks(),
+	types: {
+		pageLoad: function ( stats ) {
+			console.log( 'pageLoad init script.' );
+			// This does not need to do anything special.
+		},
+		domEvent: function ( stats ) {
+			console.log( 'pageLoad init script.' );
+			const events = {};
+			stats.forEach( function ( statToRecord ) {
+				const triggerName = statToRecord.trigger;
+				if ( statToRecord.element && statToRecord.event ) {
+					const elemToAttach = document.querySelector( statToRecord.element );
+					if ( elemToAttach && ! events[statToRecord.element] ) {
+						elemToAttach.addEventListener( statToRecord.event, function ( e ) {
+							if ( checkUniqueRecordedToday( statToRecord ) ) {
+								return;
+							}
+
+							WPJMStats.hooks.doAction( triggerName );
+						} );
+						events[statToRecord.element] = true;
+					}
+				}
+			} );
+		},
+		// New style of declaration, a stat that relies on calling a custom js func.
+		initListingImpression: function ( stats ) {
```

Comment:
I'd move all the listing impression code to its own file that exports just this function, that way it's neatly organized :)

Note the `stats` argument is not actually being used here, looks like the event is hard-coded in `logThem` above.